### PR TITLE
Define conversions and operations on at::Half

### PR DESCRIPTION
**Pull request by colesbury**
```
The goal is to be able to use at::Half throughout ATen, including in
CUDA kernels and have it operate like built-in types. This avoids the
need for cuda::from_type and cuda::to_type before every
AT_DISPATCH_ALL_TYPES_AND_HALF call.
```

### CODE REVIEW

File: aten/src/ATen/Half-inl.h [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+#ifdef __CUDACC__
+#include <cuda_fp16.h>
+#endif
+
+namespace at {
+
+/// Constructors
+
+inline AT_HOSTDEVICE Half::Half(const __half& value) {
+  std::memcpy(&x, &value, sizeof(unsigned short));
```
Comment (soumith) [inline aten/src/ATen/Half-inl.h]: they have std::memcpy in device code???

Comment (colesbury) [reply aten/src/ATen/Half-inl.h]: Yes, although it doesn't seem to be as smart about trivial memcpys so maybe adding a special case for device code makes sense. (On CPU, this ends up being equivalent to `x = value` if you compile with optimizations)

**Comment by ezyang** [comment]
Yes!!!!!

**Comment by ezyang** [comment]
CC @Jorghi12, this may interact with HIP

File: aten/src/ATen/Half-inl.h [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  static constexpr bool has_infinity = true;
+  static constexpr bool has_quiet_NaN = true;
+  static constexpr bool has_signaling_NaN = true;
+  static constexpr bool has_denorm = numeric_limits<float>::has_denorm;
+  static constexpr bool has_denorm_loss = numeric_limits<float>::has_denorm_loss;
+  static constexpr bool round_style = numeric_limits<float>::round_style;
+  static constexpr bool is_iec559 = true;
+  static constexpr bool is_bounded = true;
+  static constexpr bool is_modulo = false;
+  static constexpr bool digits = 11;
```
[the remark this answers is not in the document]
Comment (colesbury) [reply aten/src/ATen/Half-inl.h]: Hooo boy.

File: aten/src/ATen/Half-inl.h [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  static constexpr bool digits10 = 3;
+  static constexpr bool max_digits10 = 5;
+  static constexpr bool radix = 2;
+  static constexpr bool min_exponent = -13;
+  static constexpr bool min_exponent10 = -4;
+  static constexpr bool max_exponent = 16;
+  static constexpr bool max_exponent10 = 4;
+  static constexpr bool traps = false;
```
[the remark this answers is not in the document]
Comment (colesbury) [reply aten/src/ATen/Half-inl.h]: I'll inherit it from numeric_limits<float>.

File: aten/src/ATen/Half-inl.h [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+template<> class numeric_limits<at::Half> {
+ public:
+  static constexpr bool is_specialized = true;
+  static constexpr bool is_signed = true;
+  static constexpr bool is_integer = false;
+  static constexpr bool is_exact = false;
+  static constexpr bool has_infinity = true;
+  static constexpr bool has_quiet_NaN = true;
+  static constexpr bool has_signaling_NaN = true;
+  static constexpr bool has_denorm = numeric_limits<float>::has_denorm;
```
[the remark this answers is not in the document]
Comment (colesbury) [reply aten/src/ATen/Half-inl.h]: Good idea

File: aten/src/ATen/Half-inl.h [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+  static constexpr bool max_exponent10 = 4;
+  static constexpr bool traps = false;
+  static constexpr bool tinyness_before = numeric_limits<float>::tinyness_before;
+  static constexpr at::Half min() { return at::Half(0x400, at::Half::from_bits); }
+  static constexpr at::Half lowest() { return at::Half(0xFBFF, at::Half::from_bits); }
+  static constexpr at::Half max() { return at::Half(0x7BFF, at::Half::from_bits); }
+  static constexpr at::Half epsilon() { return at::Half(0x1400, at::Half::from_bits); }
+  static constexpr at::Half round_error() { return at::Half(0x3800, at::Half::from_bits); }
+  static constexpr at::Half infinity() { return at::Half(0x7C00, at::Half::from_bits); }
+  static constexpr at::Half quiet_NaN() { return at::Half(0x7E00, at::Half::from_bits); }
```
[the remark this answers is not in the document]
Comment (colesbury) [reply aten/src/ATen/Half-inl.h]: There's a lot of values for each NaN. I chose ones most similar to the versions reported by libstdc++ numeric_limits<float> (zero at every bit that doesn't matter vs. one at every bit that doesn't matter).

File: aten/src/ATen/native/cuda/EmbeddingBag.cu [hunk 6]
```diff
@@ -50,7 +46,7 @@ __global__ void EmbeddingBag_updateOutputKernel(
       int64_t end = (bag < numBags - 1) ? (offsets[bag + 1]) : numIndices;
       assert(end >= begin);
 
-      accscalar_t weightFeatSum = scalar_cast<accscalar_t>(0);
+      accscalar_t weightFeatSum = static_cast<accscalar_t>(0); // FIXME
```
[the remark this answers is not in the document]
Comment (colesbury) [reply aten/src/ATen/native/cuda/EmbeddingBag.cu]: Fixed.

File: aten/src/ATen/Half-inl.h [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+  static constexpr bool min_exponent10 = -4;
+  static constexpr bool max_exponent = 16;
+  static constexpr bool max_exponent10 = 4;
+  static constexpr bool traps = false;
+  static constexpr bool tinyness_before = numeric_limits<float>::tinyness_before;
+  static constexpr at::Half min() { return at::Half(0x400, at::Half::from_bits); }
+  static constexpr at::Half lowest() { return at::Half(0xFBFF, at::Half::from_bits); }
+  static constexpr at::Half max() { return at::Half(0x7BFF, at::Half::from_bits); }
+  static constexpr at::Half epsilon() { return at::Half(0x1400, at::Half::from_bits); }
+  static constexpr at::Half round_error() { return at::Half(0x3800, at::Half::from_bits); }
```
[the remark this answers is not in the document]
Comment (colesbury) [reply aten/src/ATen/Half-inl.h]: I don't think so - we don't emulate other rounding modes unlike the sourceforge project.

**Review by gchanan — APPROVED** [review]

**Pull request merged**